cmd/serve.go

```diff
@@ -1,17 +1,221 @@
 	"github.com/odpf/shield/config"
+	"github.com/odpf/shield/internal/group"
+	"github.com/odpf/shield/internal/org"
+	"github.com/odpf/shield/internal/project"
+	"github.com/odpf/shield/internal/roles"
+	"github.com/odpf/shield/internal/user"
+	"github.com/odpf/shield/pkg/sql"
+	"github.com/odpf/shield/proxy"
+	blobstore "github.com/odpf/shield/store/blob"
+	"github.com/odpf/shield/store/postgres"
+	"github.com/pkg/errors"
+	"github.com/pkg/profile"
 	cli "github.com/spf13/cobra"
+	"golang.org/x/net/http2"
+	"golang.org/x/net/http2/h2c"
+)
+
+var (
+	proxyTermChan         = make(chan os.Signal, 1)
+	ruleCacheRefreshDelay = time.Minute * 2
 )
 
 func serveCommand(logger log.Logger, appConfig *config.Shield) *cli.Command {
 	c := &cli.Command{
-		Use:   "serve",
-		Short: "Start server and proxy default on port 8080",
+		Use:     "serve",
+		Short:   "Start server and proxy default on port 8080",
+		Example: "shield serve",
+		RunE: func(cmd *cli.Command, args []string) error {
+			return serve(logger, appConfig)
+		},
 	}
-	c.AddCommand(proxyCommand(logger, appConfig))
-	c.AddCommand(apiCommand(logger, appConfig))
 	return c
 }
+
+func serve(logger log.Logger, appConfig *config.Shield) error {
+	if profiling := os.Getenv("SHIELD_PROFILE"); profiling == "true" || profiling == "1" {
+		defer profile.Start(profile.CPUProfile, profile.ProfilePath("."), profile.NoShutdownHook).Stop()
+	}
+
+	ctx, cancelFunc := context.WithCancel(server.HandleSignals(context.Background()))
+	defer cancelFunc()
+
+	db, dbShutdown := setupDB(appConfig.DB, logger)
+	defer dbShutdown()
+
+	var cleanUpFunc []func() error
+	var cleanUpProxies []func(ctx context.Context) error
+	cleanUpFunc, cleanUpProxies, err := startProxy(logger, appConfig, ctx, cleanUpFunc, cleanUpProxies)
+	if err != nil {
+		return err
+	}
+
+	muxServer := startServer(logger, appConfig, err, ctx, db)
+
+	waitForTermSignal(ctx)
+	cleanup(logger, ctx, cleanUpFunc, cleanUpProxies, muxServer)
+
+	return nil
+}
+
+func cleanup(logger log.Logger, ctx context.Context, cleanUpFunc []func() error, cleanUpProxies []func(ctx context.Context) error, s *server.MuxServer) {
+	for _, f := range cleanUpFunc {
+		if err := f(); err != nil {
+			logger.Warn("error occurred during shutdown", "err", err)
+		}
+	}
+	for _, f := range cleanUpProxies {
+		shutdownCtx, shutdownCancel := context.WithTimeout(ctx, time.Second*20)
+		if err := f(shutdownCtx); err != nil {
+			shutdownCancel()
+			logger.Warn("error occurred during shutdown", "err", err)
+			continue
+		}
+		shutdownCancel()
+	}
+
+	shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer shutdownCancel()
+
+	s.Shutdown(shutdownCtx)
+}
+
+func startServer(logger log.Logger, appConfig *config.Shield, err error, ctx context.Context, db *sql.SQL) *server.MuxServer {
+	s, err := server.NewMux(server.Config{
+		Port: appConfig.App.Port,
+	}, server.WithMuxGRPCServerOptions(getGRPCMiddleware(appConfig, logger)))
```

Comment:
we need to also configure zap to inject logger into ctx as well, which can be used by lower layers like grpc & store

Reply:
Injecting custom logger(wrapper over zap) into grpc would require some additional changes. will add a todo for now to push this for testing and keep track of. will try to make changes for it in background meanwhile.